pennylane/qchem/vibrational/localize_modes.py

```diff
@@ -0,0 +1,248 @@
+    qrot = _mat_transform(uparams, qmat)
+    xi_pm = np.sum(np.sum(qrot**2, axis=1) ** 2)
+
+    return -xi_pm
+
+
+def _normalize_q(qmat):
+    r"""Returns the normalized displacement vectors.
+
+    Args:
+        qmat (TensorLike[float]): matrix of displacement vectors
+
+    Returns:
+        TensorLike[float]: normalized matrix of displacement vectors
+
+    """
+
+    qnormalized = np.zeros_like(qmat)
+    nmodes = qmat.shape[2]
+
+    for m in range(nmodes):
+        m_norm = np.sum(np.abs(qmat[:, :, m]) ** 2)
+        qnormalized[:, :, m] = qmat[:, :, m] / np.sqrt(m_norm)
+
+    return qnormalized
+
+
+def _localization_unitary(qmat):
+    r"""Calculates the unitary matrix to localize the displacement vectors and
+    displacement vectors.
+
+    Args:
+        qmat (TensorLike[float]): matrix of displacement vectors associated with normal-modes
+
+    Returns:
+        tuple: A tuple containing the following:
+         - TensorLike[float] : unitary matrix to localize the displacement vectors
+         - TensorLike[float] : localized displacement vectors
+
+    """
+
+    nmodes = qmat.shape[2]
+    num_params = nmodes * (nmodes - 1) // 2
+
+    rng = qml.math.random.default_rng(1000)
+    params = 2 * np.pi * rng.random(num_params)
+
+    qnormalized = _normalize_q(qmat)
+
+    optimization_res = scipy.optimize.minimize(_params_cost, params, args=(qnormalized, nmodes))
+
+    # Check if the minimization was successful; if it wasn't, proceed with the normal modes.
+    if not optimization_res.success:
+        warnings.warn(
+            "Mode localization finished unsuccessfully, returning normal modes..."
+        )  # pragma: no cover
+        return _params_to_unitary(np.zeros_like(params), nmodes), qmat  # pragma: no cover
+
+    params_opt = optimization_res.x
+    uloc = _params_to_unitary(params_opt, nmodes)
+
+    qloc = _mat_transform(uloc, qmat)
+
+    return uloc, qloc
+
+
+def _localize_modes(freqs, vecs):
+    r"""Performs the mode localization for a given set of frequencies and displacement vectors.
+
+    Args:
+        freqs (list[float]): normal mode frequencies
+        vecs (list[float]): displacement vectors along the normal modes
+
+    Returns:
+        tuple: A tuple containing the following:
+         - list[float] : localized frequencies
+         - TensorLike[float] : localized displacement vectors
+         - TensorLike[float] : localization matrix
+
+    """
+    nmodes = len(freqs)
+    hess_normal = np.diag(np.square(freqs))
+
+    qmat = np.array([vecs[m] for m in range(nmodes)]).transpose(1, 2, 0)
+
+    uloc, qloc = _localization_unitary(qmat)
+    hess_loc = uloc.transpose() @ hess_normal @ uloc
+    loc_freqs = np.sqrt(np.array([hess_loc[m, m] for m in range(nmodes)]))
+
+    loc_perm = np.argsort(loc_freqs)
+    loc_freqs = loc_freqs[loc_perm]
+    qloc = qloc[:, :, loc_perm]
+    uloc = uloc[:, loc_perm]
+
+    return loc_freqs, qloc, uloc
+
+
+def localize_normal_modes(freqs, vecs, bins=[2600]):
```

Comment:
[optional] The name can be localize_modes.

Reply:
We already have a private function names _localize_modes, will leave this one as it is.